### DISCUSSION TITLE
Simplifying the Node hierarchy

### DISCUSSION THREAD

**Discussion by vsajip** [congo-cc/congo-parser-generator#168] [Q&A]
There are a lot of slight differences in terms of how the node hierarchy looks, depending on whether tree building is enabled. For example, whether `Token` inherits from `Node`/`BaseNode`. Why is this?  If tree building isn't enabled, then sure, we can skip generating all the tree-building code, but I'm not sure why the node hierarchy couldn't always be

```
interface Node
class BaseNode implements Node
class Token extends BaseNode
```
then, terminals can inherit from `Token` and non-terminals can inherit from `BaseNode`. Why is `Token` a completely standalone class when tree building isn't enabled? Also, why does `Token` implement `CharSequence`? It seems unintuitive at first glance ...

**Comment by revusky** [comment]
> Why is `Token` a completely standalone class when tree building isn't enabled? 

TBH, I'm a bit puzzled by the question. If tree building isn't enabled, then we don't have `Node` or `BaseNode`, so the generated `Token.java` is much more bare. Now, one can imagine that, for whatever reasons, people want to write/maintain their own tree/node API, so they turn Congo's tree-building off and then they have their own tree/node API, so they can then, if they want, have something like:

       INJECT Token : implements our.own.Node

I mean to say, if people don't want to use the Node API we generate automatically, then... Of course, it's also true that they can have tree-building turned on in Congo and still use their own custom API... maybe modulo some potential name clashes admittedly...

       INJECT Node : implements our.own.Node {....}

But maybe, for whatever reasons, somebody just wants to dispense with generating the `Node/BaseNode` that we generate... (I'm thinking out loud a bit...)

BTW, I suppose you know that currently `Token` does not extend `BaseNode`, which is taken to be the root type for non-terminal nodes. Of course, I suppose it's not inconceivable that `Token` could subclass `BaseNode`, but I'm not sure what the benefit (or drawback) of that is offhand.

I mean, the thing is that if you know roughly the history of how all this developed, you can understand where things came from. The original JavaCC, released at some point in 1996 had no automatic tree-building. Then in early 1997, JJTree was added to the package. But JJTree was just a preprocessor for JavaCC. You had a separate `jjtree` app that ran over `.jjt` files, which were just `.jj` files with some special tree-building annotations added. So, `jjtree` didn't change anything in the core JavaCC tool itself. Now, the most normal and natural thing would be that if you were using automatic tree-building, i.e. the `jjtree` tool, your tokens would extend `Node/SimpleNode`. (It was originally `SimpleNode` in JJTree, but at some point, I changed the name of `SimpleNode` to `BaseNode` because I thought it made more sense...) But anyway, retrofitting Token.java to implement `Node` never occurred to anybody in the legacy JavaCC project, despite it being the most obvious idea imaginable... (*I really think so...*)

So that's how the current situation evolved. If tree-building is *not* enabled, then we have a bare `Token.java` that does not extend `Node/BaseNode` since those classes aren't even generated! But I don't see offhand why that is a problem. I mean, granted, we could generate `Node/BaseNode` even if tree-building is explicitly turned off, but I don't quite see the point. If somebody really does just want to turn off tree-building and have something more similar to legacy JavaCC, then...

**Reply by adMartem** [reply]
The fact that `Token` does not extend `BaseNode` was an annoyance to me in the JTB extensions.  I have to wrap `Token` in an actual `NodeToken` delegate class rather than `NodeToken` just extending `Token`.  So for what it's worth, in that case there are advantages to making `Token` subclass `BaseNode`, as it would eliminate an unnecessary node simply wrapping the `Token` node.

**Reply by revusky** [reply]
Well, thinking about it again, maybe `Token` should extend `BaseNode` (IF tree-building is turned on anyway). I guess I was just explaining the history of why it is like that, not necessarily saying that it must be like that...

One thing I was just wondering about is whether an application programmer has the option of having Token extend BaseNode. I mean, supppose, in your grammar, you actually wrote:

      INJECT BASE_TOKEN_CLASS : extends NODE_PACKAGE.BASE_NODE_CLASS 

Would something break as a result? You could try it...

**Reply by adMartem** [reply]
I think I originally tried to do that (it's been awhile), and it led to inheritance ambiguity errors, but I'll try again and let you know.

**Comment by vsajip** [comment]
Not a problem as such, but we do have useful functionality such as e.g. computing a token's start and end location and image from the offsets, which would need duplicating in the source (albeit inside #ifs so only one set is generated) for the two cases. There's definitely some cognitive overhead to remember what goes where depending on grammar options, where actually it seems like it hardly matters if we always generate Node, BaseNode and Token and its subclasses but the non-terminals only when tree building.

**Reply by revusky** [reply]
Well, I think, as a practical matter, it's pretty marginal, because, well... though I could be wrong about this (but I doubt I am) tree-building being enabled is going to be, by far, the majority usage pattern. I mean, the fact is that with tree-building on, you just get so much stuff for free that... I just try to imagine some project that has some little (or not so little) grammar that they maintain and it's hard to imagine why they wouldn't just use the tree-building. I mean they just have, for free, all this API they can use to filter/traverse the nodes in a tree. Granted, it's a well understood problem, and you can write/maintain all that code yourself, but *why would you?* Well, not only do you get all this typical stuff for free, like 

       List<SomeNodeType> someNodes = rootNode.descendants(SomeNodeType.class, n->n.someConditionSatisfied());

But also you can *inject* whatever helper method you need into a Node subtype, so...

So the use of Congo with tree-building off is probably a minority of users, but if somebody does turn it off, then, well... there is no real need to generate the extra Node/BaseNode. So I would tend to leave that be.

**Comment by revusky** [comment]
> Also, why does `Token` implement `CharSequence`? It seems unintuitive at first glance ...

Well, at some point, I decided to retrofit our generated APIs so that they extended core Java APIs. So you see that `Node` extends `List<Node>` and also `TokenSource` (the superclass of XXXLexer) and `Token` both implement `java.lang.CharSequence`, which is the common API of `java.lang.String` and `StringBuilder` and some other "stringy" objects. One reason I did that is that I anticipated this being useful in various contexts. For example, you may have noticed that the generated constructors for XXXParser and XXXLexer tend to take a `CharSequence` as an argument, so I anticipate just being able to pass a `Token` in as the argument to a parser (or sub-parser to precise). Or to pass in an XXXLexer as an argument to a different sort of parser. I thought we could eventually get some extremely succinct ways of expressing things. As for this being "unnatural", well... I don't think so. I think that if you think of `CharSequence` as an interface that marks "stringy" sorts of objects, then that `Token` implements that interface makes sense. And you can directly use a `Token` in various standard Java APIs, such as:

       StringBuilder buf = new StringBuilder(someToken);

And `someToken.charAt(i)` just works without having an intermediate String object. Things like that. It certainly isn't costly to have these things implement `CharSequence`, I don't think. There's more boilerplate in terms of having `Node` extend `List<Node>` actually...

**Comment by vsajip** [comment]
Just by the way, here's an example of how having code in two places can waste time. Here's `getEndColumn()` in `Node`:
https://github.com/congo-cc/congo-parser-generator/blob/7de06d06438aee27535369adbc7aae8a5ec5e153/src/templates/java/Node.java.ftl#L393-L396
and here's the same in `Token`:
https://github.com/congo-cc/congo-parser-generator/blob/7de06d06438aee27535369adbc7aae8a5ec5e153/src/templates/java/Token.java.ftl#L263-L266
This might be the cause of #167.

**Reply by revusky** [reply]
Well, it's true that the repitition is kind of annoying. It occurs because of the no-tree-building use-case in which there is no `Node` API. So, you actually see here: https://github.com/congo-cc/congo-parser-generator/blob/7de06d06438aee27535369adbc7aae8a5ec5e153/src/templates/java/Token.java.ftl#L233 this code is only generated if tree-building is off. If tree-building is on, then Token implements Node and automatically has these methods.

**Comment by vsajip** [comment]
Also, why is the `tokenSource` property of `Token` and `BaseNode` the lexer class rather than just `TokenSource`?

**Comment by revusky** [comment]
> Also, why is the `tokenSource` property of `Token` and `BaseNode` the lexer class rather than just `TokenSource`?

Sorry to be answering this (or attempting to...) so late, but I must have just not noticed this question. It's because John Bradley wrote something else in the thread an hour ago and I just noticed this.

I guess the thing is that arguably the tokenSource property should be the `TokenSource` abstract base class rather than the `XXXLexer` concrete class, but I think I chose the latter because, in practice, in implementation, you would just end up downcasting to `XXXLexer` all over the place, so I think I finally figured I'd just make the field an instanceof `XXXLexer`. I'm not sure that it makes that much difference practically. It's almost a stylistic thing, it seems.
